llvm/lib/Target/Sparc/SparcISelLowering.cpp

```diff
@@ -1737,6 +1737,11 @@ SparcTargetLowering::SparcTargetLowering(const TargetMachine &TM,
   setOperationAction(ISD::SUBC, MVT::i32, Legal);
   setOperationAction(ISD::SUBE, MVT::i32, Legal);
 
+  if (Subtarget->isVIS3()) {
+    setOperationAction(ISD::ADDC, MVT::i64, Legal);
+    setOperationAction(ISD::ADDE, MVT::i64, Legal);
```

Comment:
It time is not a concern, I would strongly suggest to switch to legalizing UADDO/UADDO_CARRY.
ADDC/ADDE has been deprecated long ago as they effectively ban DAG optimizations.

Reply:
SG